### ISSUE TITLE
OS.execute() can't run Python

### ISSUE THREAD

**Issue by 19PHOBOSS98** [godotengine/godot#48379]
**Godot version:**
3.3.stable

**OS/device including version:**
 MacBook Pro Mojave

**Issue description:**
I tried executing this but it only ever gives me an empty array:
`
func _ready():
	var output = []
	OS.execute('python', ['--version'], true, output) 
	print(output)
`

I followed this:
https://godotforums.org/discussion/23795/cant-run-python-but-can-run-terminal-commands
and it seemed to work out with them

**Steps to reproduce:**
execute given code above

**Minimal reproduction project:**
[Terminal Test.zip](https://github.com/godotengine/godot/files/6410812/Terminal.Test.zip)

**Comment by BeayemX** [comment]
Probably requires the full path to the python binary.

On Linux this would be something like `/usr/bin/python'

**Comment by bruvzg** [comment]
Add extra `true` argument to the end of command, for some reason macOS system Python is writing version to the `stderr` (system default version is 2.7, so it might be default behavior for Python 2.x).

```gdscript
var output = []
var ret = OS.execute('python', ['--version'], true, output, true)
 print(output)
```

Python 3.9 installed via [homebrew](https://brew.sh/) output version to the `stdout`.

**Comment by 19PHOBOSS98** [comment]
> Add extra `true` argument to the end of command, for some reason macOS system Python is writing version to the `stderr` (system default version is 2.7, so it might be default behavior for Python 2.x).
> 
> ```gdscript
> var output = []
> var ret = OS.execute('python', ['--version'], true, output, true)
>  print(output)
> ```
> 
> Python 3.9 installed via [homebrew](https://brew.sh/) output version to the `stdout`.

ok, the good news is it finally printed something out, the bad news: it isn't the right thing:

https://imgur.com/undefined

**Comment by bruvzg** [comment]
App bundles won't load terminal configs (`.bash_profile`, `.zshrc`, etc.) where custom PATH is set, and will only search the default set of paths: `/usr/bin`, `/bin`, `/usr/sbin` and `/sbin`.

If you want to use custom python installation, use the full path (e.g `/usr/local/bin/python3` for homebrew installation).

> MacBook Pro Mojave

Big Sur have both python 2 (2.7.16) and python 3 (3.8.2) installed under `/usr/bin` by default, older versions probably have 2.x only.

**Comment by 19PHOBOSS98** [comment]
for the time being I just made myself a shell script instead and run python over there. My real goal was to be able to run my object detection AI model through Godot... without having to install modules. It even lets me run Conda and switch environments. So far it's not letting me down :)

My gdscript that runs the .sh file:
```
func _ready():
	var output = []
	OS.execute("sh", ["/Users/PH0B0SS/Desktop/shelltest2.sh"], true,output)
	prints(output)
```


My Crude Shell Script:
```
#!/bin/bash   
echo "Yiiiisssssss XD"

source /Users/PH0B0SS/opt/miniconda3/etc/profile.d/conda.sh
conda activate tensorflow1_15

#python /Users/PH0B0SS/Desktop/content/models-master/research/object_detection/Object_detection_image_2.py

echo $(conda --version)
echo $(python --version)
echo $(python3 --version)
```
(note: the python versions displayed depends on the python version installed on the conda environment)

It works, although it get's stuck when I try to run the "Object_detection_image_2.py". See it keeps running the python script to display the labelled image. That keeps it from ending the shell script which keeps the gdscripts "OS.execute()"  waiting until I close the image preview. Only then the shell script would continue to run and dump the echos into the gdscripts output array  printing it. 

I recommend running a separate shell script to print out the python versions first instead of waiting for that one line of code to step off.

Although I can bet a dime there should be a more efficient way in running terminal commands through Godot instead of running a gdscript to run a shell script to a run python code.

**Comment by Calinou** [comment]
> It works, although it get's stuck when I try to run the "Object_detection_image_2.py". See it keeps running the python script to display the labelled image. That keeps it from ending the shell script which keeps the gdscripts "OS.execute()" waiting until I close the image preview. Only then the shell script would continue to run and dump the echos into the gdscripts output array printing it.

Set the environment variable `PYTHONUNBUFFERED` to `1` before running the process. In 3.3, you can do this using `OS.set_environment()` *before* using `OS.execute()`.

**Comment by 19PHOBOSS98** [comment]
Update

> App bundles won't load terminal configs (`.bash_profile`, `.zshrc`, etc.) where custom PATH is set, and will only search the default set of paths: `/usr/bin`, `/bin`, `/usr/sbin` and `/sbin`.
> 
> If you want to use custom python installation, use the full path (e.g `/usr/local/bin/python3` for homebrew installation).
> 
> > MacBook Pro Mojave
> 
> Big Sur have both python 2 (2.7.16) and python 3 (3.8.2) installed under `/usr/bin` by default, older versions probably have 2.x only.

the code works:
`var ret = OS.execute('/usr/local/bin/python3', ['--version'], true, output, true)`
shows the right version too.
The odd thing is that I took a peek under the '/usr/local/bin/' folder and I found that I don't have python 2 over there but only python3 and I'm running Mojave which is an older version.

After running my code a few more times, I noticed that Godot's output of python3's version from running the '.sh' file  through `OS.execute()`wasn't the same as running it directly from the terminal app:
https://imgur.com/a/YGDVFkb

You might think that it was Godot's fault, but nah it was Terminal's for a change.

after slapping my brain on some stuff I found out that python3 starts acting funny when I try to ask for its version each time I start up. See, my terminal normally starts up in conda's "(base)" environment or at least I thought it did. I ran the code:
`python3 --version` right off the bat. it returned `Python 3.7.0`. 

From 'base' I ran `conda activate tensorflow1_15` to open up my tensorflow environment, right. I tried running `python3 --version` again for this environment and it gave back the same thing: `Python 3.7.0`. 

I stepped down to the root environment: 
```
(tensorflow1_15) PH0B0SSs-MacBook-Pro:~ PH0B0SS$ conda deactivate
(base) PH0B0SSs-MacBook-Pro:~ PH0B0SS$ conda deactivate
```

then I went to ask what version of python3 was installed in the root:
`PH0B0SSs-MacBook-Pro:~ PH0B0SS$ python3 --version`

and you would never guess what came out:
`Python 3.7.0`

there was no way everyone was using 3.7.0 cause I know there was 3.7.10 somewhere. So I tried and went back up:
`conda activate base` (I also tried using `conda activate` seems like the same thing)
and asked it again for its python3 version number and low and behold it was being honest for a change and said:
`Python 3.8.5`
Went back down to root and asked it again for its python3 version:
`Python 3.7.0`
I opened my tensorflow1_15 environment and also checked its python3 version:
`Python 3.7.10`

So, I started a fresh run of Terminal and started off by deactivating conda to get to the root directory and run the .sh file:
```
(base) PH0B0SSs-MacBook-Pro:~ PH0B0SS$ conda deactivate
PH0B0SSs-MacBook-Pro:~ PH0B0SS$ sh /Users/PH0B0SS/Desktop/shelltest.sh
Python 2.7.10

Python 3.7.0
Yiiiisssssss XD
conda 4.9.2
Python 3.7.10
Python 3.7.10
```
It's either something wrong with python3 or my terminal but whatever it is it wasn't Godot's fault.
Turns out Godot was telling the truth about the versions and was running the shell script from root as intended.
It had trouble with the first `echo $(python3 --version)` command tho. ¯\\\_(ツ)\_/¯
